### PR TITLE
refactor: references to use specific ref type

### PR DESCRIPTION
**Pull request by strideynet**

### CODE REVIEW

**Comment by strideynet** [comment]
Address #96

File: core/virtual_machine_build.go [hunk 1]
```diff
@@ -43,7 +43,7 @@ type virtualMachineBuildCreateRequest struct {
 	Hostname            string                 `json:"hostname,omitempty"`
 	Organization        *Organization          `json:"organization,omitempty"`
 	Zone                *Zone                  `json:"zone,omitempty"`
-	DataCenter          *DataCenter            `json:"data_center,omitempty"`
+	DataCenter          DataCenterRef          `json:"data_center,omitempty"`
```
[the remark this answers is not in the document]
Comment (strideynet) [reply core/virtual_machine_build.go:46]: I'll agree on removing the omitempty here.

File: core/data_center_test.go [hunk 2]
```diff
@@ -282,137 +220,6 @@ func TestDataCentersClient_List(t *testing.T) {
 	}
 }
 
-func TestDataCentersClient_Get(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (strideynet) [reply core/data_center_test.go:285]: Valid point.

File: core/data_center_test.go [hunk 3]
```diff
@@ -57,112 +56,51 @@ func TestDataCenter_JSONMarshaling(t *testing.T) {
 	}
 }
 
-func TestNewDataCenterLookup(t *testing.T) {
-	type args struct {
-		idOrPermalink string
-	}
+func TestNewDataCenter_Ref(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (strideynet) [reply core/data_center_test.go:59]: Agreed

File: core/data_center.go [hunk 4]
```diff
@@ -16,33 +14,27 @@ type DataCenter struct {
 	Country   *Country `json:"country,omitempty"`
 }
 
-// NewDataCenterLookup takes a string that is a DataCenter ID or Permalink,
-// returning a empty *DataCenter struct with either the ID or Permalink field
-// populated with the given value. This struct is suitable as input to other
-// methods which accept a *DataCenter as input.
-func NewDataCenterLookup(
```
[the remark this answers is not in the document]
Comment (strideynet) [reply core/data_center.go:23]: Hmm, I'm almost never a fan of ambiguity (unless theres a really strong use-case argument).

I think I'd want to answer the following questions:
- How likely are we to stick with IDs prefixed with some kind of entity description?
- How much work would arise if we were to change the IDs?
- What's the size of the potential risk to users running an older version of the library, with incompatible IDs
- How much value are we actually providing by allowing ambiguity here? And what happens if a passed in string cannot be identified as one of two (or three) options?

[the remark this answers is not in the document]
Comment (jimeh) [reply core/data_center.go:23]: > - How likely are we to stick with IDs prefixed with some kind of entity description?

At this point, I'd say VERY likely. It's baked in quite deep in Katapult.

> - How much work would arise if we were to change the IDs?

Would depend on how we're changing them. If we'd be simply dropping the prefix value and keeping a random alphanumeric string, probably a lot. But if we were to switch to UUIDv4 for example, it's simple enough to verify if a string is a valid UUIDv4.

> - What's the size of the potential risk to users running an older version of the library, with incompatible IDs

Unknown, but in theory we probably wouldn't change the IDs without a new major version of the API, meaning old client code would keep working until the old API is deprecated.

> - How much value are we actually providing by allowing ambiguity here? And what happens if a passed in string cannot be identified as one of two (or three) options?

I'm leaning more and more towards it not providing much value, and in fact probably encouraging ambiguity. And I'm not sure my previous thinking of us officially providing the functionality to prevent external parties from making assumptions about our ID formats is that valid. Users who need this indeterministic form of lookup would most likely be better off not relying on a magic lookup, and instead simply trying to lookup by permalink (or relevant alt-field) first, and if that fails fall back on a lookup by ID.

TL;DR: I think we can remove the ability to turn a ambiguous lookup string into a explicitly defined lookup field/value combo.

**Review by jimeh — CHANGES_REQUESTED** [review]

File: core/data_center_test.go [hunk 5]
```diff
@@ -57,112 +56,61 @@ func TestDataCenter_JSONMarshaling(t *testing.T) {
...
 		},
 	}
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			got := tt.obj.lookupReference()
 
-			assert.Equal(t, tt.want, got)
+	for _, tt := range tests {
+		t.Run(t.Name(), func(t *testing.T) {
+			assert.Equal(t, tt.want, tt.ref.queryValues())
```
[the remark this answers is not in the document]
Comment (strideynet) [reply core/data_center_test.go:113]: I think I agree here, switching to an inline want url.Values is probably going to be much clearer.

File: core/virtual_machine_network_interface.go [hunk 6]
```diff
@@ -48,18 +45,18 @@ type virtualMachineNetworkInterfacesResponseBody struct {
...
-	AddressVersion                 IPVersion                       `json:"address_version,omitempty"`
+	VirtualMachineNetworkInterface VirtualMachineNetworkInterfaceRef `json:"virtual_machine_network_interface,omitempty"`
+	AddressVersion                 IPVersion                         `json:"address_version,omitempty"`
 }
 
 type virtualMachineNetworkInterfaceUpdateSpeedProfileRequest struct {
-	VirtualMachineNetworkInterface *VirtualMachineNetworkInterface `json:"virtual_machine_network_interface,omitempty"`
-	SpeedProfile                   *NetworkSpeedProfile            `json:"speed_profile,omitempty"`
+	VirtualMachineNetworkInterface VirtualMachineNetworkInterfaceRef `json:"virtual_machine_network_interface,omitempty"`
+	SpeedProfile                   NetworkSpeedProfileRef            `json:"speed_profile,omitempty"`
```
[the remark this answers is not in the document]
Comment (strideynet) [reply core/virtual_machine_network_interface.go:59]: Yup. That's a good idea.

File: core/data_center_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (strideynet) [reply core/data_center_test.go:113]: As agreed in conversation, we will stick with testQueryableEncoding() and fix this at a later date.

File: core/virtual_machine_build.go [hunk 7]
```diff
@@ -38,24 +44,24 @@ const (
...
 
 type virtualMachineBuildCreateRequest struct {
 	Hostname            string                   `json:"hostname,omitempty"`
 	Organization        OrganizationRef          `json:"organization"`
-	Zone                ZoneRef                  `json:"zone,omitempty"`
-	DataCenter          DataCenterRef            `json:"data_center"`
-	Package             VirtualMachinePackageRef `json:"package,omitempty"`
-	DiskTemplate        DiskTemplateRef          `json:"disk_template,omitempty"`
+	Zone                *ZoneRef                 `json:"zone,omitempty"`
+	DataCenter          *DataCenterRef           `json:"data_center"`
```
Comment (jimeh) [inline core/virtual_machine_build.go:60]: I think a `omitempty` flag is missing here :)

```suggestion
	DataCenter          *DataCenterRef           `json:"data_center,omitempty"`
```

File: core/data_center_test.go [hunk 8]
```diff
@@ -57,112 +56,51 @@ func TestDataCenter_JSONMarshaling(t *testing.T) {
...
-			want: &DataCenter{},
 		},
 	}
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			got := tt.obj.lookupReference()
 
-			assert.Equal(t, tt.want, got)
+	for _, tt := range tests {
+		t.Run(t.Name(), func(t *testing.T) {
```
Comment (jimeh) [inline core/data_center_test.go:102]: Small fix to use the test case name in the sub-test.

```suggestion
		t.Run(tt.name, func(t *testing.T) {
```

**Review by jimeh — CHANGES_REQUESTED** [review]
Only two minor tweaks are needed I think, along with re-generating golden files so we get rid of old left-over golden files for which the tests no longer exist. This can be done with:

```bash
make regen-golden
```

**Review by jimeh — APPROVED** [review]
👍
